File: exercise5/python/task2.py
import numpy as np
import matplotlib.pyplot as plt
from common1 import *
from common2 import *

edge_threshold = 0 # todo: choose an appropriate value
blur_sigma     = 0 # todo: choose an appropriate value
filename       = '../data/image1_und.jpg'

I_rgb      = plt.imread(filename)
I_rgb      = I_rgb/255.0
I_gray     = rgb2gray(I_rgb)
I_blur     = blur(I_gray, blur_sigma)
Iu, Iv, Im = central_difference(I_blur)
u,v,theta  = extract_edges(Iu, Iv, Im, edge_threshold)

#
# Task 2a: Compute accumulator array H
#
bins      = 100 # Placeholder
rho_max   = 0 # Placeholder
rho_min   = 0 # Placeholder
theta_min = 0 # Placeholder
theta_max = 0 # Placeholder
H = np.zeros([bins,bins]) # Placeholder

# Tip: Use histogram2d for task 2a
# H, _, _ = np.histogram2d(theta, rho, bins=bins, range=[[theta_min, theta_max], [rho_min, rho_max]])
# H = H.T # Make rows be rho and columns be theta (see documentation)

#
# Task 2b: Find local maxima
#
line_threshold = 15 # Placeholder
window_size = 10 # Placeholder
peak_rows,peak_cols = extract_peaks(H, window_size, line_threshold)

#
# Task 2c: Convert peak (row, column) pairs into (theta, rho) pairs.
#
peak_theta = [0, 0.2, 0.5, 0.7] # Placeholder to demonstrate use of draw_line
peak_rho   = [10, 100, 300, 500] # Placeholder to demonstrate use of draw_line

plt.figure(figsize=[6,8])
plt.subplot(211)
plt.imshow(H, extent=[theta_min, theta_max, rho_min, rho_max], aspect='auto')
plt.xlabel('$\\theta$ (radians)')
plt.ylabel('$\\rho$ (pixels)')
plt.colorbar(label='Votes')
plt.title('Hough transform histogram')
plt.subplot(212)
plt.imshow(I_rgb)
plt.xlim([0, I_rgb.shape[1]])
plt.ylim([I_rgb.shape[0], 0])
for i in range(len(peak_theta)):
    draw_line(peak_theta[i], peak_rho[i], color='yellow')
plt.tight_layout()
plt.savefig('out2.png')
# plt.show()

File: exercise5/matlab/task2.m
clear;
edge_threshold = 0.02; % todo: choose an appropriate value
blur_sigma     = 1; % todo: choose an appropriate value
line_threshold = 10; % todo: choose an appropriate value
bins           = 500; % todo: choose an appropriate value
filename       = '../data/image1_und.jpg';

I_rgb       = imread(filename);
I_rgb       = im2double(I_rgb);
I_gray      = rgb2gray(I_rgb);
I_blur      = blur(I_gray, blur_sigma);
[Iu,Iv,Im]  = central_difference(I_blur);
[u,v,theta] = extract_edges(Iu, Iv, Im, edge_threshold);

%
% Task 2a: Compute accumulator array H
%
rho_max   = norm(size(I_rgb)); % Placeholder
rho_min   = -rho_max; % Placeholder
theta_min = -pi; % Placeholder
theta_max = pi; % Placeholder


rho = u.*cos(theta) + v.*sin(theta); 
[H, ~, ~] = histcounts2(theta, rho, bins, ...
    'XBinLimits', [theta_min, theta_max], ...
    'YBinLimits', [rho_min, rho_max]);
H = H'; % Make rows be rho and columns be theta

window_size = 11;
[peak_rows,peak_cols] = extract_peaks(H, window_size, line_threshold);

%
% Task 2c: Convert peak (row, column) pairs into (theta, rho) pairs.
%
peak_rho = rho_min + (rho_max - rho_min)*(peak_rows-0.5)/bins;
peak_theta = theta_min + (theta_max - theta_min)*(peak_cols-0.5)/bins;

subplot(211);
imagesc(H, 'XData', [theta_min theta_max], 'YData', [rho_min rho_max]);
xlabel('\theta (radians)');
ylabel('\rho (pixels)');
cb = colorbar();
cb.Label.String = 'Votes';
subplot(212);
imshow(I_rgb); hold on;
for i=1:size(peak_rho)
    draw_line(peak_theta(i), peak_rho(i));
end
